File: sdk/src/agent/builder.rs
use crate::agent::consumer::{
    AgentHandler, AgentMiddleware, ConcurrencyPolicy, DeadLetterSink, Deduplicator,
    ReliableConsumer, RetryPolicy,
};
use crate::agent::router::InboxRoute;
use crate::error::LaserError;
use crate::laser::Laser;
use crate::provenance::AgentTopic;
use crate::types::{AgentId, ConsumerGroupName};
use laser_wire::agent::{AgentCard, CapabilityDescriptor};
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::oneshot;
use tokio::task::JoinHandle;
use tracing::warn;

/// One logical agent defaults to one Iggy consumer group.
///
/// `spawn` derives the group name from `id` unless `consumer_group` is set. Horizontal
/// scaling (multiple replicas of the same agent) means multiple processes
/// joining that group, and Iggy load-balances messages across them.
///
/// **Do NOT put two different agent ids in the same group.** The
/// `ReliableConsumer` skips messages whose `agdx.to` does not
/// match its own `id`, which is the right behavior when a publisher
/// addresses a specific agent on a shared topic. If two agents shared
/// a group, Iggy would still deliver each message to exactly one member,
/// and if it landed on the wrong one the filter would drop it and the
/// intended agent would never see it. Each `Agent::builder().id(...).spawn(..)`
/// call is its own consumer group. Do not bypass that by reusing the
/// consumer group name across distinct agent ids.
#[derive(bon::Builder)]
pub struct Agent<H> {
    /// The agent's stable logical identity.
    pub id: AgentId,
    /// Deployment group for load balancing replicas. Defaults to the agent id's
    /// spelling but remains a distinct type and may be overridden explicitly.
    pub consumer_group: Option<ConsumerGroupName>,
    /// The topic this agent consumes.
    pub listen_on: AgentTopic<'static>,
    /// The handler invoked for each message.
    pub handler: H,
    /// Default reply topic for `AgentCtx::respond` (none = `respond` errors).
    pub respond_on: Option<AgentTopic<'static>>,
    /// Default route from a resolved target agent to the topic its work is sent
    /// on, used by [`AgentCtx::fan_out`](crate::agent::AgentCtx::fan_out) and the
    /// directed-send helpers. Sticks as this agent's routing convention. Defaults
    /// to [`InboxRoute::Advertised`] (resolve each target to its live-presence
    /// inbox). Set [`InboxRoute::Fixed`] for a workflow topic the caller owns.
    #[builder(default)]
    pub inbox_route: InboxRoute,
    /// Override the consumer poll interval (default: the runtime's reactive 10ms).
    pub poll_interval: Option<Duration>,
    /// How long [`AgentHandle::shutdown`] waits for the in-flight message to
    /// finish draining before the consumer is dropped. [`AgentHandle::abort`] is
    /// the unconditional hard stop. Defaults to the consumer's 30s grace.
    pub shutdown_grace: Option<Duration>,
    /// How message handling is scheduled across the partitions this agent is
    /// assigned. Defaults to [`ConcurrencyPolicy::Serial`]. Set
    /// [`ConcurrencyPolicy::SerialPerPartition`] so a slow or retrying message on
    /// one partition does not stall the others.
    pub concurrency: Option<ConcurrencyPolicy>,
    #[builder(default)]
    pub warm_dedup: bool,
    /// Cross-cutting hooks wrapped around each handler dispatch, in order, for
    /// auth, metrics, and tracing without touching the handler.
    #[builder(default)]
    pub middleware: Vec<Arc<dyn AgentMiddleware>>,
    /// Notified on every dead-letter with the result of publishing it, so a lost
    /// poison message is an observable event rather than only a log line.
    pub on_dead_letter: Option<Arc<dyn DeadLetterSink>>,
    /// Override the dedup window size (recent idempotency keys kept in memory).
    /// Defaults to the [`ReliableConsumer`] default when unset.
    pub dedup_window: Option<usize>,
    /// Override the handler retry policy (attempts + backoff for a retryable
    /// handler error). Defaults to the [`RetryPolicy`] default when unset.
    pub retry: Option<RetryPolicy>,
    /// Override the dedup backend (e.g. a durable `StateStore`-backed one). Defaults
    /// to the in-memory sliding window.
    pub deduplicator: Option<Box<dyn Deduplicator>>,
    /// The signature verifier for control and effect topics: when set, every
    /// message's envelope signature is verified against this registry before
    /// dispatch, and an unsigned or unverified record is dead-lettered. This is
    /// the enforcement gate for authorship and authorization, so it is reachable
    /// from the primary builder rather than only the lower-level consumer.
    #[cfg(feature = "sign")]
    pub verifier: Option<std::sync::Arc<crate::sign::KeyRegistry>>,
    /// This agent's signing identity: when set, `AgentCtx::respond` answers a
    /// correlated command with a signed AGDX response instead of a plain reply,
    /// so a caller holding a verifier (and the contract path's signer binding)
    /// can accept this agent's terminals and no one else's.
    #[cfg(feature = "sign")]
    pub signing_key: Option<std::sync::Arc<crate::sign::SigningKey>>,
    /// What this agent serves. When non-empty, on spawn the agent publishes a
    /// capability card to the registry (so capability routing can discover it) and
    /// advertises its inbox (its `listen_on` topic) as live presence (so a fan-out
    /// resolves where to send it work). Empty means a private agent that does not
    /// advertise, the prior behavior.
    #[builder(default)]
    pub capabilities: Vec<CapabilityDescriptor>,
    /// Emit a `Working` status the moment an AGDX command is picked up, before the
    /// handler runs, so a [`Laser::contract`](crate::laser::Laser::contract) caller
    /// can tell the command was consumed. Off by default. Enable it on agents that
    /// take contracted work, with a `respond_on` set.
    #[builder(default)]
    pub ack_on_pickup: bool,
    /// Run `consolidator` every `consolidate_every` off the handler loop (its
    /// own task, stopped with the agent). Both must be set for the tick to
    /// exist. Absent means no background consolidation, ever. The scope is the
    /// consolidator's own concern (a `DefaultConsolidator` holds its memory).
    pub consolidate_every: Option<Duration>,
    /// The consolidation pass the periodic tick runs (see
    /// [`consolidate_every`](Self::consolidate_every)).
    pub consolidator: Option<crate::memory::SharedConsolidator>,
    /// The pre-effect policy hook applied to everything this agent's handler
    /// publishes (`ctx.send`/`respond`/`request`/`fan_out`, AGDX verbs, memory
    /// writes), with the [`GovernorMode`](crate::govern::GovernorMode) it runs
    /// under. Spawn re-scopes the agent's `Laser` with it, so a governor set on
    /// the connection is replaced for this agent. See
    /// [`Laser::with_governor`](crate::laser::Laser::with_governor).
    pub governor: Option<(
        Arc<dyn crate::govern::ActionGovernor>,
        crate::govern::GovernorMode,
    )>,
}

impl<H> Agent<H>
where
    H: AgentHandler + Sync + Send + 'static,
{
    /// Spawn the agent on `laser` as its own consumer group, returning a handle to stop it.
    #[tracing::instrument(target = "laser", level = "info", skip_all, fields(agent = %self.id, operation = "spawn"))]
    pub fn spawn(self, laser: Laser) -> AgentHandle {
        let laser = match self.governor {
            Some((governor, mode)) => laser.with_governor(governor, mode),
            None => laser,
        };
        let id = self.id.clone();
        let listen_on = self.listen_on.clone();
        let capabilities = self.capabilities;
        let group = self
            .consumer_group
            .unwrap_or_else(|| ConsumerGroupName::for_agent(&id));
        let topic = self.listen_on.topic_string();
        let handler = self.handler;
        let respond_on = self.respond_on;
        let inbox_route = self.inbox_route;
        let ack_on_pickup = self.ack_on_pickup;
        let poll_interval = self.poll_interval;
        let warm_dedup = self.warm_dedup;
        let deduplicator = self.deduplicator;
        let shutdown_grace = self.shutdown_grace;
        let concurrency = self.concurrency;
        let dedup_window = self.dedup_window;
        let retry = self.retry;
        let middleware = self.middleware;
        let on_dead_letter = self.on_dead_letter;
        #[cfg(feature = "sign")]
        let verifier = self.verifier;
        #[cfg(feature = "sign")]
        let signing_key = self.signing_key;
        let (shutdown, shutdown_rx) = oneshot::channel();
        let (ready_tx, ready_rx) = oneshot::channel();
        // The consolidation tick, off the handler loop: best-effort, logged,
        // aborted with the agent. No background magic unless both knobs are set.
        let consolidation = match (self.consolidate_every, self.consolidator) {
            (Some(every), Some(consolidator)) => Some(tokio::spawn(async move {
                let mut tick = tokio::time::interval(every);
                tick.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
                loop {
                    tick.tick().await;
                    let scope = crate::memory::MemoryScope::default();
                    if let Err(error) =
                        crate::memory::Consolidator::consolidate(&consolidator, &scope).await
                    {
                        warn!(%error, "background consolidation pass failed");
                    }
                }
            })),
            _ => None,
        };
        let task = tokio::spawn(async move {
            if !capabilities.is_empty() {
                advertise(&laser, &id, &listen_on, capabilities).await?;
            }
            let consumer = ReliableConsumer::builder()
                .group(group)
                .agent(id)
                .topic(topic)
                .maybe_respond_on(respond_on)
                .inbox_route(inbox_route)
                .ack_on_pickup(ack_on_pickup)
                .maybe_poll_interval(poll_interval)
                .maybe_shutdown_grace(shutdown_grace)
                .maybe_concurrency(concurrency)
                .warm_dedup(warm_dedup)
                .maybe_deduplicator(deduplicator)
                .maybe_dedup_window(dedup_window)
                .maybe_retry(retry)
                .middleware(middleware)
                .maybe_on_dead_letter(on_dead_letter);
            #[cfg(feature = "sign")]
            let consumer = consumer
                .maybe_verifier(verifier)
                .maybe_signing_key(signing_key);
            consumer
                .build()
                .run(&laser, handler, ready_tx, shutdown_rx)
                .await
        });
        AgentHandle {
            shutdown,
            task,
            ready: Some(ready_rx),
            consolidation,
        }
    }
}

/// Advertise a capability-bearing agent before it starts consuming: publish a
/// non-expiring capability card to the registry (durable discovery), and, where
/// the server serves the presence command, advertise its inbox (its `listen_on`
/// topic) as live presence (so capability fan-out resolves where to send work).
/// Both are best-effort: a failure logs and the agent still runs, since a missing
/// advertisement degrades discovery but must not stop the agent.
pub(crate) async fn advertise(
    laser: &Laser,
    id: &AgentId,
    listen_on: &AgentTopic<'static>,
    capabilities: Vec<CapabilityDescriptor>,
) -> Result<(), LaserError> {
    let card = AgentCard {
        name: None,
        version: None,
        capabilities,
        ttl_micros: None,
    };
    if let Err(error) = laser.publish_card(id.clone(), &card).await {
        warn!(%error, agent = %id, "failed to publish the capability card");
    }
    #[cfg(feature = "query")]
    {
        let presence = laser_wire::agent::AgentPresence::new(id.wire_id())
            .with_inbox(listen_on.topic_string());
        if let Err(error) = laser.advertise_presence(&presence).await {
            if matches!(error, LaserError::PresenceConflict { .. }) {
                return Err(error);
            }
            // Expected against a streaming server without the presence command,
            // where routing falls back to a fixed inbox route. Not an error.
            tracing::debug!(%error, agent = %id, "inbox presence not advertised");
        }
    }
    #[cfg(not(feature = "query"))]
    let _ = listen_on;
    Ok(())
}

/// Owns a spawned agent. Dropping it detaches the task and leaves it running, as
/// before. Call `shutdown` (or `join`) to stop it and observe a consumer error.
pub struct AgentHandle {
    shutdown: oneshot::Sender<()>,
    task: JoinHandle<Result<(), LaserError>>,
    ready: Option<oneshot::Receiver<()>>,
    // The background consolidation tick, aborted whenever the agent stops.
    consolidation: Option<JoinHandle<()>>,
}

impl AgentHandle {
    /// Wait until the agent has joined its group and is polling, so a publish after
    /// this is delivered rather than racing the join. Idempotent.
    pub async fn ready(&mut self) -> Result<(), LaserError> {
        if let Some(ready) = self.ready.take() {
            ready.await.map_err(|_| {
                LaserError::HandlerConfig("agent stopped before it became ready".to_owned())
            })?;
        }
        Ok(())
    }

    /// Signal the agent to stop, wait for it, and surface any consumer error.
    pub async fn shutdown(self) -> Result<(), LaserError> {
        let _ = self.shutdown.send(());
        if let Some(consolidation) = &self.consolidation {
            consolidation.abort();
        }
        Self::join_task(self.task).await
    }

    /// Wait for the agent to finish (it runs until its consumer ends or errors).
    pub async fn join(self) -> Result<(), LaserError> {
        let result = Self::join_task(self.task).await;
        if let Some(consolidation) = &self.consolidation {
            consolidation.abort();
        }
        result
    }

    /// Abort the agent's task immediately, without waiting.
    pub fn abort(&self) {
        self.task.abort();
        if let Some(consolidation) = &self.consolidation {
            consolidation.abort();
        }
    }

    async fn join_task(task: JoinHandle<Result<(), LaserError>>) -> Result<(), LaserError> {
        match task.await {
            Ok(result) => result,
            Err(join) => Err(LaserError::HandlerConfig(join.to_string())),
        }
    }
}

File: foreign/typescript/src/agent/builder.ts
import { HandlerConfigError, TimeoutError } from "../client/errors.js"
import type { Laser } from "../client/laser.js"
import { ConsumerGroupName, type AgentId } from "../types/ids.js"
import type { ActionGovernor, GovernorMode } from "../govern.js"
import type { KeyRegistry, SigningKey } from "../signing.js"
import type { CapabilityDescriptor } from "../wire/agent.js"
import type { MemoryScope } from "../memory/types.js"
import {
  ReliableConsumer,
  type AgentHandler,
  type AgentMiddleware,
  type ConcurrencyPolicy,
  type DeadLetterSink,
  type Deduplicator,
  type RetryPolicy
} from "./reliable-consumer.js"
import type { InboxRoute } from "./router.js"

interface AgentDefinition {
  readonly id: AgentId
  readonly consumerGroup?: ConsumerGroupName
  readonly listenOn: string
  readonly handler: AgentHandler
  readonly respondOn?: string
  readonly inboxRoute?: InboxRoute
  readonly pollIntervalMs?: number
  readonly shutdownGraceMs: number
  readonly concurrency?: ConcurrencyPolicy
  readonly warmDedup: boolean
  readonly middleware: readonly AgentMiddleware[]
  readonly deadLetterSink?: DeadLetterSink
  readonly dedupWindow?: number
  readonly retry?: RetryPolicy
  readonly deduplicator?: Deduplicator
  readonly capabilities: readonly CapabilityDescriptor[]
  readonly ackOnPickup: boolean
  readonly verifier?: KeyRegistry
  readonly signingKey?: SigningKey
  readonly governor?: ActionGovernor
  readonly governorMode?: GovernorMode
  readonly consolidateEveryMs?: number
  readonly consolidator?: AgentConsolidator
}

export interface AgentConsolidator {
  consolidate(scope: MemoryScope): Promise<unknown>
}

type RunOutcome = { readonly kind: "ok" } | { readonly kind: "error"; readonly error: unknown }

/** Controls one running agent and drains it on async disposal. */
export class AgentHandle implements AsyncDisposable {
  private readonly shutdownController = new AbortController()
  private readonly hardStopController = new AbortController()
  private hardStopped = false
  private readySettled = false
  private readonly readyPromise: Promise<void>
  private readonly resolveReady: () => void
  private readonly rejectReady: (error: unknown) => void
  private readonly task: Promise<RunOutcome>
  private readonly shutdownGraceMs: number
  private readonly consolidationController = new AbortController()
  private readonly consolidationTask: Promise<void> | undefined

  constructor(definition: AgentDefinition, laser: Laser) {
    this.shutdownGraceMs = definition.shutdownGraceMs
    let resolveReady: (() => void) | undefined
    let rejectReady: ((error: unknown) => void) | undefined
    this.readyPromise = new Promise<void>((resolve, reject) => {
      resolveReady = resolve
      rejectReady = reject
    })
    this.resolveReady = resolveReady as () => void
    this.rejectReady = rejectReady as (error: unknown) => void
    this.task = this.run(definition, laser)
    this.consolidationTask =
      definition.consolidateEveryMs !== undefined && definition.consolidator !== undefined
        ? this.runConsolidation(
            definition.consolidateEveryMs,
            definition.consolidator,
            this.consolidationController.signal
          )
        : undefined
  }

  ready(): Promise<void> {
    return this.readyPromise
  }

  /** Stops intake, drains active handlers, and joins background work. */
  async shutdown(): Promise<void> {
    this.shutdownController.abort("agent shutdown")
    this.consolidationController.abort("agent shutdown")
    let timer: ReturnType<typeof setTimeout> | undefined
    const timeout = new Promise<"timeout">((resolve) => {
      timer = setTimeout(() => {
        resolve("timeout")
      }, this.shutdownGraceMs)
    })
    const outcome = await Promise.race([this.task, timeout])
    if (timer !== undefined) clearTimeout(timer)
    if (outcome === "timeout") {
      this.hardStopped = true
      throw new TimeoutError("agent shutdown drain")
    }
    if (outcome.kind === "error") throw outcome.error
    await this.consolidationTask
  }

  /** Delegates async disposal to `shutdown()`. */
  [Symbol.asyncDispose](): Promise<void> {
    return this.shutdown()
  }

  async join(): Promise<void> {
    const outcome = await this.task
    this.consolidationController.abort("agent stopped")
    await this.consolidationTask
    if (outcome.kind === "error") throw outcome.error
  }

  abort(): void {
    this.hardStopped = true
    this.hardStopController.abort("agent abort")
    this.shutdownController.abort("agent abort")
    this.consolidationController.abort("agent abort")
  }

  private async runConsolidation(
    everyMs: number,
    consolidator: AgentConsolidator,
    signal: AbortSignal
  ): Promise<void> {
    while (!signal.aborted) {
      try {
        await consolidator.consolidate({})
      } catch {
        // Consolidation is best effort and must not stop message handling.
      }
      await this.waitForConsolidationInterval(everyMs, signal)
    }
  }

  private waitForConsolidationInterval(everyMs: number, signal: AbortSignal): Promise<void> {
    if (signal.aborted) return Promise.resolve()
    return new Promise<void>((resolve) => {
      const onAbort = (): void => {
        clearTimeout(timer)
        resolve()
      }
      const timer = setTimeout(() => {
        signal.removeEventListener("abort", onAbort)
        resolve()
      }, everyMs)
      signal.addEventListener("abort", onAbort, { once: true })
    })
  }

  private async run(definition: AgentDefinition, laser: Laser): Promise<RunOutcome> {
    try {
      const scoped =
        definition.governor !== undefined && definition.governorMode !== undefined
          ? laser.withGovernor(definition.governor, definition.governorMode)
          : laser
      if (definition.capabilities.length > 0) {
        await scoped.agent(definition.id).advertise(definition.listenOn, definition.capabilities)
      }
      const consumer = new ReliableConsumer({
        group: definition.consumerGroup ?? ConsumerGroupName.forAgent(definition.id),
        topic: definition.listenOn,
        agent: definition.id,
        ...(definition.respondOn !== undefined ? { respondOn: definition.respondOn } : {}),
        ...(definition.inboxRoute !== undefined ? { inboxRoute: definition.inboxRoute } : {}),
        ...(definition.pollIntervalMs !== undefined
          ? { pollIntervalMs: definition.pollIntervalMs }
          : {}),
        ...(definition.concurrency !== undefined ? { concurrency: definition.concurrency } : {}),
        ...(definition.deadLetterSink !== undefined
          ? { deadLetterSink: definition.deadLetterSink }
          : {}),
        ...(definition.dedupWindow !== undefined ? { dedupWindow: definition.dedupWindow } : {}),
        ...(definition.retry !== undefined ? { retry: definition.retry } : {}),
        ...(definition.deduplicator !== undefined ? { deduplicator: definition.deduplicator } : {}),
        warmDedup: definition.warmDedup,
        middleware: definition.middleware,
        ackOnPickup: definition.ackOnPickup,
        ...(definition.verifier !== undefined ? { verifier: definition.verifier } : {}),
        ...(definition.signingKey !== undefined ? { signingKey: definition.signingKey } : {})
      })
      await consumer.run(scoped, definition.handler, {
        signal: this.shutdownController.signal,
        hardSignal: this.hardStopController.signal,
        hardAborted: () => this.hardStopped,
        ready: () => {
          this.readySettled = true
          this.resolveReady()
        }
      })
      if (!this.readySettled) this.resolveReady()
      return { kind: "ok" }
    } catch (error) {
      if (!this.readySettled) {
        this.readySettled = true
        this.rejectReady(error)
      }
      return { kind: "error", error }
    }
  }
}

export class AgentBuilder {
  private agentId: AgentId | undefined
  private group: ConsumerGroupName | undefined
  private listenTopic: string | undefined
  private replyTopic: string | undefined
  private agentHandler: AgentHandler | undefined
  private route: InboxRoute | undefined
  private pollMs: number | undefined
  private graceMs = 30_000
  private concurrencyPolicy: ConcurrencyPolicy | undefined
  private shouldWarmDedup = false
  private readonly middlewareList: AgentMiddleware[] = []
  private sink: DeadLetterSink | undefined
  private window: number | undefined
  private retryPolicy: RetryPolicy | undefined
  private dedup: Deduplicator | undefined
  private advertisedCapabilities: readonly CapabilityDescriptor[] = []
  private pickupAck = false
  private signatureVerifier: KeyRegistry | undefined
  private signer: SigningKey | undefined
  private actionGovernor: ActionGovernor | undefined
  private mode: GovernorMode | undefined
  private consolidationIntervalMs: number | undefined
  private memoryConsolidator: AgentConsolidator | undefined

  id(id: AgentId): this {
    this.agentId = id
    return this
  }

  consumerGroup(group: ConsumerGroupName): this {
    this.group = group
    return this
  }

  listenOn(topic: string): this {
    this.listenTopic = topic
    return this
  }

  respondOn(topic: string): this {
    this.replyTopic = topic
    return this
  }

  handler(handler: AgentHandler): this {
    this.agentHandler = handler
    return this
  }

  inboxRoute(route: InboxRoute): this {
    this.route = route
    return this
  }

  pollInterval(ms: number): this {
    this.pollMs = ms
    return this
  }

  shutdownGrace(ms: number): this {
    this.graceMs = ms
    return this
  }

  concurrency(policy: ConcurrencyPolicy): this {
    this.concurrencyPolicy = policy
    return this
  }

  warmDedup(value = true): this {
    this.shouldWarmDedup = value
    return this
  }

  middleware(middleware: AgentMiddleware): this {
    this.middlewareList.push(middleware)
    return this
  }

  deadLetterSink(sink: DeadLetterSink): this {
    this.sink = sink
    return this
  }

  dedupWindow(size: number): this {
    this.window = size
    return this
  }

  retry(retry: RetryPolicy): this {
    this.retryPolicy = retry
    return this
  }

  deduplicator(deduplicator: Deduplicator): this {
    this.dedup = deduplicator
    return this
  }

  capabilities(capabilities: readonly CapabilityDescriptor[]): this {
    this.advertisedCapabilities = capabilities
    return this
  }

  ackOnPickup(value = true): this {
    this.pickupAck = value
    return this
  }

  verifier(verifier: KeyRegistry): this {
    this.signatureVerifier = verifier
    return this
  }

  signingKey(signingKey: SigningKey): this {
    this.signer = signingKey
    return this
  }

  governor(governor: ActionGovernor, mode: GovernorMode): this {
    this.actionGovernor = governor
    this.mode = mode
    return this
  }

  consolidateEvery(milliseconds: number): this {
    this.consolidationIntervalMs = milliseconds
    return this
  }

  consolidator(consolidator: AgentConsolidator): this {
    this.memoryConsolidator = consolidator
    return this
  }

  spawn(laser: Laser): AgentHandle {
    if (this.agentId === undefined) throw new HandlerConfigError("Agent.builder().id() is required")
    if (this.listenTopic === undefined) {
      throw new HandlerConfigError("Agent.builder().listenOn() is required")
    }
    if (this.agentHandler === undefined) {
      throw new HandlerConfigError("Agent.builder().handler() is required")
    }
    if (!Number.isFinite(this.graceMs) || this.graceMs < 0) {
      throw new HandlerConfigError("shutdownGrace must be a non-negative finite number")
    }
    if (this.pollMs !== undefined && (!Number.isFinite(this.pollMs) || this.pollMs < 0)) {
      throw new HandlerConfigError("pollInterval must be a non-negative finite number")
    }
    if (this.window !== undefined && (!Number.isSafeInteger(this.window) || this.window < 1)) {
      throw new HandlerConfigError("dedupWindow must be a positive safe integer")
    }
    if (
      this.retryPolicy !== undefined &&
      (!Number.isSafeInteger(this.retryPolicy.maxAttempts) || this.retryPolicy.maxAttempts < 1)
    ) {
      throw new HandlerConfigError("retry maxAttempts must be a positive safe integer")
    }
    if (
      this.consolidationIntervalMs !== undefined &&
      (!Number.isFinite(this.consolidationIntervalMs) || this.consolidationIntervalMs <= 0)
    ) {
      throw new HandlerConfigError("consolidateEvery must be a positive finite number")
    }
    if (
      this.retryPolicy !== undefined &&
      (!Number.isFinite(this.retryPolicy.baseDelayMs) || this.retryPolicy.baseDelayMs < 0)
    ) {
      throw new HandlerConfigError("retry baseDelayMs must be a non-negative finite number")
    }
    if (
      this.concurrencyPolicy?.kind === "serial-per-partition" &&
      (!Number.isSafeInteger(this.concurrencyPolicy.maxPartitions) ||
        this.concurrencyPolicy.maxPartitions < 1)
    ) {
      throw new HandlerConfigError("maxPartitions must be a positive safe integer")
    }
    return new AgentHandle(
      {
        id: this.agentId,
        ...(this.group !== undefined ? { consumerGroup: this.group } : {}),
        listenOn: this.listenTopic,
        handler: this.agentHandler,
        ...(this.replyTopic !== undefined ? { respondOn: this.replyTopic } : {}),
        ...(this.route !== undefined ? { inboxRoute: this.route } : {}),
        ...(this.pollMs !== undefined ? { pollIntervalMs: this.pollMs } : {}),
        shutdownGraceMs: this.graceMs,
        ...(this.concurrencyPolicy !== undefined ? { concurrency: this.concurrencyPolicy } : {}),
        warmDedup: this.shouldWarmDedup,
        middleware: [...this.middlewareList],
        ...(this.sink !== undefined ? { deadLetterSink: this.sink } : {}),
        ...(this.window !== undefined ? { dedupWindow: this.window } : {}),
        ...(this.retryPolicy !== undefined ? { retry: this.retryPolicy } : {}),
        ...(this.dedup !== undefined ? { deduplicator: this.dedup } : {}),
        capabilities: this.advertisedCapabilities,
        ackOnPickup: this.pickupAck,
        ...(this.signatureVerifier !== undefined ? { verifier: this.signatureVerifier } : {}),
        ...(this.signer !== undefined ? { signingKey: this.signer } : {}),
        ...(this.actionGovernor !== undefined ? { governor: this.actionGovernor } : {}),
        ...(this.mode !== undefined ? { governorMode: this.mode } : {}),
        ...(this.consolidationIntervalMs !== undefined
          ? { consolidateEveryMs: this.consolidationIntervalMs }
          : {}),
        ...(this.memoryConsolidator !== undefined ? { consolidator: this.memoryConsolidator } : {})
      },
      laser
    )
  }
}

export const Agent = {
  builder(): AgentBuilder {
    return new AgentBuilder()
  }
} as const
